### PR TITLE
Debugger - Breakpoints and Variable Explorer

### PR DESCRIPTION
**Pull request by gushuro**
Adding breakpoints inside blocks
Adding debug mode decorations to editor.
Added functionality to the variable explorer (see https://github.com/Microsoft/pxt/pull/5263 for previous discussion on this).

debugger v1 0

### CODE REVIEW

File: webapp/src/app.tsx [hunk 1]
```diff
@@ -2129,8 +2131,17 @@ export class ProjectView
 
     toggleDebugging() {
         const state = !this.state.debugging;
+        console.log("turning debugging mode to " + state);
         this.setState({ debugging: state, tracing: false });
+        var blocks = this.blocksEditor.editor.getAllBlocks();
```
Comment (riknoll) [inline webapp/src/app.tsx]: Use `let` not `var` (irony!)

File: webapp/src/blocks.tsx [hunk 2]
```diff
@@ -28,11 +29,49 @@ export class Editor extends toolboxeditor.ToolboxEditor {
     showCategories: boolean = true;
     filters: pxt.editor.ProjectFilters;
     showSearch: boolean;
+    breakpointsByBlock: Map<string, number>; // Map block id --> breakpoint ID
```
Comment (riknoll) [inline webapp/src/blocks.tsx]: Use `pxt.Map<number>`

File: webapp/src/debugger.tsx [hunk 3]
```diff
@@ -139,8 +174,14 @@ export class DebuggerVariables extends data.Component<DebuggerVariablesProps, De
 
         return Object.keys(variables).length == 0 ? <div /> :
             <div className={`ui segment debugvariables ${frozen ? "frozen" : ""}`}>
-                <div className="ui middle aligned list">
-                    {this.renderVariables(variables)}
+                <div className="ui collapsing celled table">
+                <thead>
+                    <tr><th>Variable</th>
```
Comment (riknoll) [inline webapp/src/debugger.tsx]: Nitpick: Don't split tags like this, either put them all on the same line or give the open and close tags their own line.

File: webapp/src/debugger.tsx [hunk 4]
```diff
@@ -112,24 +113,58 @@ export class DebuggerVariables extends data.Component<DebuggerVariablesProps, De
...
-                <div role="listitem" className={`ui label image variable ${v.prevValue !== undefined ? "changed" : ""}`} style={{ backgroundColor: varcolor }}
-                    onClick={onClick}>
-                    <span className="varname">{variable}</span>
-                    <div className="detail">
-                        <span className="varval">{DebuggerVariables.renderValue(v.value)}</span>
-                        <span className="previousval">{v.prevValue !== undefined ? `${DebuggerVariables.renderValue(v.prevValue)}` : ''}</span>
-                    </div>
-                </div>
-            </div>);
+            const onMouseOver : any = undefined; // prob a bad idea. Maybe onMouseEnter --> onMouseLeave?
```
Comment (riknoll) [inline webapp/src/debugger.tsx]: Just pass an empty function if you aren't using it.

File: theme/debugger.less [hunk 5]
```diff
@@ -97,6 +113,20 @@ div.simframe div.pause-overlay {
...
+}
+
+.debugger #downloadArea {
+    background: @mainMenuDebuggerBackground;
+    //background-color: lightgreen
+}
+
+.debugger #filelist {
+    background: @mainMenuDebuggerBackground;
+    background-color: lightgreen
```
Comment (riknoll) [inline theme/debugger.less]: Put this `lightgreen` in a variable

File: theme/debugger.less [hunk 6]
```diff
@@ -63,17 +63,33 @@
     }
 
     .variable {
-        color: white;
+    }
+    .transparent{
+        opacity: 0;
     }
     .variable.changed {
-        outline: solid 2px yellow;
```
Comment (riknoll) [inline theme/debugger.less]: Just delete the empty less rules; vscode complains about them

File: webapp/src/debugger.tsx (hunk 3 above)
Comment (pelikhan) [inline webapp/src/debugger.tsx]: Use lf macro

File: webapp/src/debugger.tsx [hunk 7]
```diff
@@ -139,8 +174,14 @@ export class DebuggerVariables extends data.Component<DebuggerVariablesProps, De
 
         return Object.keys(variables).length == 0 ? <div /> :
             <div className={`ui segment debugvariables ${frozen ? "frozen" : ""}`}>
-                <div className="ui middle aligned list">
-                    {this.renderVariables(variables)}
+                <div className="ui collapsing celled table">
+                <thead>
+                    <tr><th>Variable</th>
+                    <th>Type/Value</th></tr>
```
Comment (pelikhan) [inline webapp/src/debugger.tsx]: Use lf macro

File: theme/debugger.less (hunk 6 above)
Comment (gushuro) [reply theme/debugger.less]: Sure.

File: pxtsim/debugger.ts [hunk 8]
```diff
@@ -119,6 +132,17 @@ namespace pxsim {
                     r[k.replace(/___\d+$/, '')] = valToJSON(frame[k])
                 }
             }
+            if (frame.fields) {
+                // Fields of an object.
+                for (let k of Object.keys(frame.fields)) {
+                    r[k.replace(/___\d+$/, '')] = valToJSON(frame.fields[k])
+                }
+            } else if (frame.data) {
```
[the remark this answers is not in the document]
Comment (gushuro) [reply pxtsim/debugger.ts]: Where?

[the remark this answers is not in the document]
Comment (gushuro) [reply pxtsim/debugger.ts]: Alright Richard explained offline. Thanks to both.

File: webapp/src/debugger.tsx [hunk 9]
```diff
@@ -112,35 +114,81 @@ export class DebuggerVariables extends data.Component<DebuggerVariablesProps, De
...
+            let type = this.variableType(v);
             const onClick = v.value && v.value.id ? () => this.toggle(v) : undefined;
-            r.push(<div key={(parent || "") + variable} className="item">
-                <div role="listitem" className={`ui label image variable ${v.prevValue !== undefined ? "changed" : ""}`} style={{ backgroundColor: varcolor }}
-                    onClick={onClick}>
-                    <span className="varname">{variable}</span>
+
+            r.push(<tr key={(parent || "") + variable} className="item" style={{padding: "0em",}}>
+                <td className={`variable ${v.prevValue !== undefined ? "changed" : ""}`}
+                    onClick={onClick} onMouseOver={this.onMouseOverVariable} style={{ padding: 0.2, }}>
```
Comment (riknoll) [inline webapp/src/debugger.tsx]: Seems like this passing should go in the css

**Comment by gushuro** [comment]
We'll leave further improvements on the variable explorer for future PRs.
- onClick and onHover performance issues.
- Refactor expand/collapse logic
- Move variable explorer to Toolbox

**Pull request merged**